File: Top-Interview-Questions-Leetcode/Medium-Collections/array-and-strings/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        isPalin = [[False for _ in range(n)] for _ in range(n)]
        ans = 1
        
        for i in range(n):
            isPalin[i][i] = True
            
        for l in range(2, n + 1):
            for i in range(0, n - l + 1):
                j = i + l - 1
                if l == 2:
                    isPalin[i][j] = (s[i] == s[j])
                else:
                    isPalin[i][j] = (s[i] == s[j]) & isPalin[i + 1][j - 1]
                
                ans = max(ans, l) if isPalin[i][j] else ans
        
        for i in range(n):
            if isPalin[i][i + ans - 1]:
                return s[i:i + ans]

File: Top-Interview-Questions-Leetcode/Medium-Collections/array-and-strings/longest-palindromic-substring.cpp
#include <iostream>

class Solution {
public:
    std::string longestPalindrome(std::string s) {
        int n = s.size(), l = 0, h = 0;
        
        for (int i = 0, left, right; i < n - 1; ++i) {
            left = i, right = i + 1;
            while (left >= 0 and right < n and s[left] == s[right]) {
                if (right - left + 1 > h - l + 1) l = left, h = right;
                left--, right++;
            }
            
            left = i - 1, right = i + 1;
            while (left >= 0 and right < n and s[left] == s[right]) {
                if (right - left + 1 > h - l + 1) l = left, h = right;
                left--, right++;
            }
            
        }
        
        return s.substr(l, h - l + 1);
    }
};